hs_tools_resource/utils.py

```diff
@@ -58,3 +62,48 @@ def get_SupportedSharingStatus_choices():
             ['Discoverable', 'Discoverable'],
             ['Private', 'Private'],
             ]
+
+
+def do_work_when_launching_app_as_needed(app_shortkey, res_shortkey, user):
+    """
+    check whether there are extra work needed to be done when the app is launched.
+    :param app_shortkey: shortkey of the app tool resource to be launched
+    :param res_shortkey: shortkey of the resource launching the app tool resource
+    :param user: requesting user
+    :return: error message if an exception is raised; otherwise, return None
+    """
+
+    # When app resource has iRODS federation target path and target resource defined as
+    # extended metadata, the resource needs to be pushed to the specified iRODS federation path
+    # in the specified iRODS storage resource.
+    try:
+        res = get_resource_by_shortkey(res_shortkey)
+        app_res = get_resource_by_shortkey(app_shortkey)
+
+        # check whether irods_path_key and irods_resc_key are added as extended metadata of the
+        # app tool resource, and if they are, pass them out as context variables for app launcher
+        # to push resource to federated iRODS path accordingly
+        irods_path_key = 'irods_federation_target_path'
+        irods_resc_key = 'irods_federation_target_resource'
+        filterd_app_obj = BaseResource.objects.filter(short_id=app_shortkey).filter(
+            extra_metadata__has_key=irods_path_key).filter(
+            extra_metadata__has_key=irods_resc_key).first()
+        if filterd_app_obj:
+            if not user.is_authenticated():
```

Comment:
The user should already be authenticated before getting this far. I believe this may be removed.

Reply:
@sblack-usu Thanks for this comment since it revealed an issue of the current implementation. Specifically, if a user is not authenticated, the current implementation does not call this utility function. However, the app tool will be launched since the app tool should be launchable for a public resource, which would result in exception from external web apps since external web apps require user authentication. Ideally it would work best if it were app-configuration depending on whether the app requires user authentication when being launched. I think I should make it so since I can imagine some apps would allow an unauthenticated user to launch the app, in that case, ```anonymous``` would be the user name. I will make changes accordingly then to add an extra configuration flag in web app tool extended metadata, and acts accordingly. Let me know if you have other thoughts.